repos/system_upgrade/common/actors/addupgradebootentry/libraries/addupgradebootentry.py

```diff
@@ -114,3 +209,101 @@ def fix_grub_config_error(conf_file, error_type):
 
     elif error_type == 'missing newline':
         write_to_file(conf_file, config + '\n')
+
+
+def local_os_stat(path):
+    """ Local wrapper around os.stat so we can safely mock it in tests. """
+    return os.stat(path)
+
+
+def _get_device_uuid(path):
+    """
+    Find the UUID of a device in which the given path is located.
+    """
+    while not os.path.ismount(path):
+        path = os.path.dirname(path)
+
+    needle_dev_id = local_os_stat(path).st_dev
+
+    for uuid in os.listdir('/dev/disk/by-uuid'):
+        uuid_fullpath = os.path.join('/dev/disk/by-uuid/', uuid)
+        dev_path = os.readlink(uuid_fullpath)
+
+        # The link target is likely relative to the UUID_fullpath, e.g., ../../dm-1.
+        # Joining it will '/dev/disk/by-uuid' will resolve the relative path.
+        # If dev_path is absolute it returns dev_path.
+        dev_path = os.path.join('/dev/disk/by-uuid', dev_path)
+        dev_path = os.path.abspath(dev_path)
+
+        dev_id = local_os_stat(dev_path).st_rdev
+        if dev_id == needle_dev_id:
+            return uuid
+
+    return None
+
+
+def _get_rdlvm_args():
+    # should we not check args returned by grubby instead?
```

Comment:
@abadger @MichalHe I've been thinking about that and I do not think so. It's coming from the original Chris work and I guess he was thinking about the freshness of the data. Which in this case we do not consider grubby output at this moment as Source of Truth (SoT) - not because of some grubby bugs at all, but due to nature that  what is written in discovered configuration file does not have to reflect reality (I've seen that couple of times) and also from the same point why we use /proc/cmdline instead of grubby at the very initial scan of the system - because we want to work with values with which we know the system booted successfully. If user changes the args without the reboot and start leapp to upgrade, it's from our pov basically invalid already (there are couple of reasons why, but keeping my answer shorter).